lib/Optimizer/CodeGen/LowerToQIR.cpp

```diff
@@ -441,24 +441,66 @@ class OneTargetRewrite : public ConvertOpToLLVMPattern<OP> {
       return success();
     }
 
-    // Get symbol for
-    // void invokeWithControlQubits(const std::size_t nControls, void
-    // (*QISFunction)(Array*, Qubit*), Qubit*, ...);
-    auto applyMultiControlFunction =
-        cudaq::opt::factory::createLLVMFunctionSymbol(
-            cudaq::opt::NVQIRInvokeWithControlBits,
-            LLVM::LLVMVoidType::get(context),
-            {i64Type, LLVM::LLVMPointerType::get(instOpQISFunctionType)},
-            parentModule, true);
-
+    // Check if all controls are qubit types, if so
+    // retain existing functionality
+    auto allControlsAreQubits = [&](auto controls) {
+      for (auto c : controls)
+        if (c.getType() != qirQubitPointerType)
+          return false;
+      return true;
+    }(adaptor.getControls());
```

Comment:
For my own benefit, why are we passing in `adaptor.getControls()` when we're already capturing everything by reference? It seems like this would be adding an extra copy (but maybe the compiler is smart enough to get rid of it).

Reply:
I think you are right. I can just drop the lambda argument here and use adaptor directly.